File: apps/api/app/schemas/community.py
"""社区 Pydantic schemas (Sprint 6 BE-S6-006/007/008/009).

请求 / 响应模型对齐 spec/13 §主线 D + alembic 0014_community.

设计要点:
- ``content`` 字段 server-side 长度限制 (post=500, comment=200) 用 Pydantic Field
  约束; 业务层不重复校验
- 列表 / 详情区分: 列表轻量 (不带评论列表), 详情含 ``user_nickname`` join
- 状态 / 分类 / 拒绝原因等都走 ``Literal`` 严格枚举
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

PostStatus = Literal["pending", "published", "rejected", "deleted", "hidden"]
PostVisibility = Literal["public", "self_only"]
PostCategory = Literal["general", "ipo_discuss", "experience"]
CommentStatus = Literal["pending", "published", "rejected", "deleted"]
LikeTargetType = Literal["post", "comment"]
ReportReason = Literal[
    "spam", "illegal", "misleading", "privacy", "pornographic", "other"
]
ReportStatus = Literal["pending", "resolved", "dismissed"]
RejectionReason = Literal[
    "content_violation", "privacy_leak", "spam", "other"
]


# ─── Post ──────────────────────────────────────────────────────────────


class PostCreateRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=500, description="帖子内容")
    category: PostCategory = "general"
    related_ipo_code: str | None = Field(default=None, max_length=16)


class PostDetailResponse(BaseModel):
    """帖子详情 (含 user 冗余字段, 列表 / 详情共用)."""

    id: uuid.UUID
    user_id: uuid.UUID
    user_nickname: str | None = None
    user_avatar_url: str | None = None
    content: str
    status: PostStatus
    visibility: PostVisibility
    category: PostCategory
    related_ipo_code: str | None = None
    likes_count: int
    comments_count: int
    reports_count: int
    rejection_reason: RejectionReason | None = None
    is_liked: bool = Field(default=False, description="当前用户是否已赞")
    created_at: datetime
    updated_at: datetime


class PostListResponse(BaseModel):
    items: list[PostDetailResponse]
    total: int
    page: int
    page_size: int


# ─── Sprint 11 BE-S11-C02: admin 管理 schemas ─────────────────────────


class AdminPostListItem(BaseModel):
    """admin 视角帖子条目. 比 PostDetailResponse 少 is_liked, 多 reviewed_by/at."""

    id: uuid.UUID
    user_id: uuid.UUID
    user_nickname: str | None = None
    user_avatar_url: str | None = None
    content: str
    status: PostStatus
    visibility: PostVisibility
    category: PostCategory
    related_ipo_code: str | None = None
    likes_count: int
    comments_count: int
    reports_count: int
    rejection_reason: RejectionReason | str | None = None
    reviewed_by: uuid.UUID | None = None
    reviewed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class AdminPostListResponse(BaseModel):
    items: list[AdminPostListItem]
    total: int
    page: int
    page_size: int


class AdminPostStatusUpdate(BaseModel):
    """``PATCH /admin/community/posts/{id}/status``."""

    status: PostStatus = Field(
        ..., description="published / pending / rejected / deleted / hidden"
    )
    reason: str | None = Field(
        default=None,
        max_length=200,
        description="处理原因 (admin 留痕迹用; 写入 rejection_reason 字段)",
    )


class AdminPostVisibilityUpdate(BaseModel):
    """``PATCH /admin/community/posts/{id}/visibility``."""

    visibility: PostVisibility = Field(
        ...,
        description="public / self_only; self_only 等同于'软隐藏', 不动 status",
    )


# ─── Comment ──────────────────────────────────────────────────────────


class CommentCreateRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=200)
    parent_comment_id: uuid.UUID | None = Field(
        default=None, description="二级评论的父评论 id; 为空 = 一级评论"
    )


class CommentResponse(BaseModel):
    id: uuid.UUID
    post_id: uuid.UUID
    user_id: uuid.UUID
    user_nickname: str | None = None
    user_avatar_url: str | None = None
    parent_comment_id: uuid.UUID | None = None
    content: str
    status: CommentStatus
    likes_count: int
    is_liked: bool = False
    created_at: datetime


class CommentListResponse(BaseModel):
    items: list[CommentResponse]
    total: int


# ─── Like ──────────────────────────────────────────────────────────────


class LikeRequest(BaseModel):
    target_type: LikeTargetType
    target_id: uuid.UUID


class LikeResponse(BaseModel):
    target_type: LikeTargetType
    target_id: uuid.UUID
    liked: bool
    likes_count: int


# ─── Report ────────────────────────────────────────────────────────────


class ReportRequest(BaseModel):
    target_type: LikeTargetType
    target_id: uuid.UUID
    reason: ReportReason
    detail: str | None = Field(default=None, max_length=500)


class ReportResponse(BaseModel):
    id: uuid.UUID
    target_type: LikeTargetType
    target_id: uuid.UUID
    reason: ReportReason
    status: ReportStatus
    created_at: datetime

File: apps/mp/api/community.ts
/**
 * 社区 API 客户端 (FE-S6-005/006/007 接 BE-S6-006/007).
 *
 * 对齐后端契约 (apps/api/app/schemas/community.py + apps/api/app/api/v1/community.py).
 *
 * 端点权限:
 * - GET /community/posts (匿名可读, FE 用 ``skipAuth: true``)
 * - GET /community/posts/{id} (匿名可读, ``skipAuth: true``)
 * - GET /community/posts/{id}/comments (匿名可读, ``skipAuth: true``)
 * - 其它 (POST 帖 / 评 / 赞 / 举报 / DELETE) 都强制登录
 *
 * 错误码 (parseCommunityError):
 * - 401 → unauthorized "请先登录"
 * - 403 → forbidden / 内容违规 / 新用户 7d 只读
 * - 404 → not_found "帖子或评论不存在"
 * - 429 → too_many_requests "操作过于频繁"
 * - 422 → invalid_field
 */

import { APIError, request } from '@/utils/request'

export type PostStatus = 'pending' | 'published' | 'rejected' | 'deleted' | 'hidden'
export type PostVisibility = 'public' | 'self_only'
export type PostCategory = 'general' | 'ipo_discuss' | 'experience'
export type CommentStatus = 'pending' | 'published' | 'rejected' | 'deleted'
export type LikeTargetType = 'post' | 'comment'
export type ReportReason =
  | 'spam'
  | 'illegal'
  | 'misleading'
  | 'privacy'
  | 'pornographic'
  | 'other'
export type RejectionReason =
  | 'content_violation'
  | 'privacy_leak'
  | 'spam'
  | 'other'

// ─── Post ───────────────────────────────────────────────────────────

export interface PostDetail {
  id: string
  user_id: string
  user_nickname: string | null
  user_avatar_url: string | null
  content: string
  status: PostStatus
  visibility: PostVisibility
  category: PostCategory
  related_ipo_code: string | null
  likes_count: number
  comments_count: number
  reports_count: number
  rejection_reason: RejectionReason | null
  is_liked: boolean
  /** ISO8601 */
  created_at: string
  updated_at: string
}

export interface PostListResponse {
  items: PostDetail[]
  total: number
  page: number
  page_size: number
}

export interface PostCreateRequest {
  content: string
  category?: PostCategory
  related_ipo_code?: string
}

export interface ListPostsParams {
  category?: PostCategory
  related_ipo_code?: string
  user_id?: string
  page?: number
  page_size?: number
}

export function listPosts(params: ListPostsParams = {}) {
  // 不用 URLSearchParams: 微信小程序 JSCore 没暴露这个全局 (H5/App 才有);
  // 同款规避见 ``api/ipo.ts:fetchIPOList``.
  const data: Record<string, string | number> = {}
  if (params.category) data.category = params.category
  if (params.related_ipo_code) data.related_ipo_code = params.related_ipo_code
  if (params.user_id) data.user_id = params.user_id
  if (params.page !== undefined) data.page = params.page
  if (params.page_size !== undefined) data.page_size = params.page_size
  return request<PostListResponse>({
    url: '/api/v1/community/posts',
    method: 'GET',
    data,
    skipAuth: true,
  })
}

export function getPost(postId: string) {
  return request<PostDetail>({
    url: `/api/v1/community/posts/${postId}`,
    method: 'GET',
    skipAuth: true,
  })
}

export function createPost(req: PostCreateRequest) {
  return request<PostDetail>({
    url: '/api/v1/community/posts',
    method: 'POST',
    data: req,
  })
}

export function deletePost(postId: string) {
  return request<void>({
    url: `/api/v1/community/posts/${postId}`,
    method: 'DELETE',
  })
}

// ─── Comment ───────────────────────────────────────────────────────

export interface CommentDetail {
  id: string
  post_id: string
  user_id: string
  user_nickname: string | null
  user_avatar_url: string | null
  parent_comment_id: string | null
  content: string
  status: CommentStatus
  likes_count: number
  is_liked: boolean
  created_at: string
}

export interface CommentListResponse {
  items: CommentDetail[]
  total: number
}

export interface CommentCreateRequest {
  content: string
  parent_comment_id?: string
}

export interface ListCommentsParams {
  parent_comment_id?: string
  page?: number
  page_size?: number
}

export function listComments(postId: string, params: ListCommentsParams = {}) {
  // 不用 URLSearchParams: 微信小程序 JSCore 没暴露这个全局 (H5/App 才有);
  // 同款规避见 ``api/ipo.ts:fetchIPOList``.
  const data: Record<string, string | number> = {}
  if (params.parent_comment_id) data.parent_comment_id = params.parent_comment_id
  if (params.page !== undefined) data.page = params.page
  if (params.page_size !== undefined) data.page_size = params.page_size
  return request<CommentListResponse>({
    url: `/api/v1/community/posts/${postId}/comments`,
    method: 'GET',
    data,
    skipAuth: true,
  })
}

export function createComment(postId: string, req: CommentCreateRequest) {
  return request<CommentDetail>({
    url: `/api/v1/community/posts/${postId}/comments`,
    method: 'POST',
    data: req,
  })
}

export function deleteComment(commentId: string) {
  return request<void>({
    url: `/api/v1/community/comments/${commentId}`,
    method: 'DELETE',
  })
}

// ─── Like ──────────────────────────────────────────────────────────

export interface LikeResponse {
  target_type: LikeTargetType
  target_id: string
  liked: boolean
  likes_count: number
}

export function toggleLike(target_type: LikeTargetType, target_id: string) {
  return request<LikeResponse>({
    url: '/api/v1/community/likes',
    method: 'POST',
    data: { target_type, target_id },
  })
}

// ─── Report ────────────────────────────────────────────────────────

export interface ReportRequest {
  target_type: LikeTargetType
  target_id: string
  reason: ReportReason
  detail?: string
}

export interface ReportResponse {
  id: string
  target_type: LikeTargetType
  target_id: string
  reason: ReportReason
  status: 'pending' | 'resolved' | 'dismissed'
  created_at: string
}

export function createReport(req: ReportRequest) {
  return request<ReportResponse>({
    url: '/api/v1/community/reports',
    method: 'POST',
    data: req,
  })
}

// ─── 错误处理 ──────────────────────────────────────────────────────

export function parseCommunityError(err: unknown): { code: string; message: string } {
  if (err instanceof APIError) {
    if (err.statusCode === 401) return { code: 'unauthorized', message: '请先登录' }
    if (err.statusCode === 403) {
      const msg = err.message || ''
      if (msg.includes('新用户')) {
        return { code: 'new_user_readonly', message: '新用户 7 天内不能发布内容' }
      }
      if (msg.includes('违规') || msg.includes('内容')) {
        return { code: 'content_violation', message: msg }
      }
      return { code: 'forbidden', message: msg || '无权操作' }
    }
    if (err.statusCode === 404) return { code: 'not_found', message: '内容不存在或已删除' }
    if (err.statusCode === 422) return { code: 'invalid_field', message: err.message ?? '字段校验失败' }
    if (err.statusCode === 429) return { code: 'too_many_requests', message: '操作过于频繁, 请稍后再试' }
    return { code: 'unknown', message: err.message ?? '网络错误' }
  }
  return { code: 'unknown', message: (err as Error)?.message ?? '网络错误' }
}
